packages/jaeger-ui/src/model/ddg/Graph.test.js

```diff
@@ -0,0 +1,234 @@
+
+      // Validate that there is an edge connecting the vertex with each expected focalSideNeighbor
+      expectedEdgeCount += focalSideNeighbors.length;
+      const focalSideEdges = Array.from(
+        new Set(pathElems.map(elem => graph.pathElemToEdge.get(elem)))
+      ).filter(Boolean);
+      const focalSideKeys = focalSideEdges.map(({ to, from }) => (to === vertex.key ? from : to));
+      const expectedKeys = focalSideNeighbors.map(
+        idx => graph.pathElemToVertex.get(graph.visIdxToPathElem[idx]).key
+      );
+      expect(focalSideKeys).toEqual(expectedKeys);
+    });
+
+    // Validate that there aren't any rogue vertices nor edges
+    expect(graph.vertices.size).toBe(expectedVertices.length);
+    expect(new Set(graph.pathElemToEdge.values()).size).toBe(expectedEdgeCount);
+  }
+
+  describe('getVertexKey', () => {
+    const testFocalElem = simpleModel.paths[0].members[2];
+    const expectedKeyEntry = pathElem => `${pathElem.operation.service.name}\t${pathElem.operation.name}`;
+    const expectedFocalElemKey = expectedKeyEntry(testFocalElem);
+    // Because getVertexKey is completely context-unaware until late-alpha, an empty ddg is sufficient to test
+    // this method.
+    const emptyGraph = new Graph({ ddgModel: { visIdxToPathElem: [] } });
+
+    it('creates key for focal pathElem', () => {
+      expect(emptyGraph.getVertexKey(testFocalElem)).toBe(expectedFocalElemKey);
+    });
+
+    it('creates key for an upstream pathElem', () => {
+      const targetElem = simpleModel.paths[0].members[0];
+      const interimElem = simpleModel.paths[0].members[1];
+      expect(emptyGraph.getVertexKey(targetElem)).toBe(
+        [expectedKeyEntry(targetElem), expectedKeyEntry(interimElem), expectedFocalElemKey].join('\n')
+      );
+    });
+
+    it('creates key for a downstream pathElem', () => {
+      const targetElem = simpleModel.paths[0].members[4];
+      const interimElem = simpleModel.paths[0].members[3];
+      expect(emptyGraph.getVertexKey(targetElem)).toBe(
+        [expectedFocalElemKey, expectedKeyEntry(interimElem), expectedKeyEntry(targetElem)].join('\n')
+      );
+    });
+  });
+
+  describe('constructor', () => {
+    it('creates five vertices and four edges for one-path ddg', () => {
+      const testGraph = new Graph({
+        ddgModel: simpleModel,
+      });
+      validateGraph(testGraph, [
+        {
+          visIndices: [0],
+        },
+        {
+          visIndices: [1],
+          focalSideNeighbors: [0],
+        },
+        {
+          visIndices: [2],
+          focalSideNeighbors: [0],
+        },
+        {
+          visIndices: [3],
+          focalSideNeighbors: [1],
+        },
+        {
+          visIndices: [4],
+          focalSideNeighbors: [2],
+        },
+      ]);
+    });
+  });
+
+  describe('convergent paths', () => {
+    it('adds separate vertices for equal PathElems that have different focalPaths, even those with equal focalSideNeighbors', () => {
```

Comment:
What is a `focalPaths`?

Reply:
discussed IRL, but for the future, `focalPath` is the DDG equivalent to `ancestralPath`. Ancestral Path is based off of the trace's root span, while `focalPath` is based off of the focalNode